sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,892 @@
+    X,
+    algo,
+    min_samples=5,
+    alpha=1.0,
+    metric="euclidean",
+    leaf_size=40,
+    n_jobs=None,
+    **metric_params,
+):
+    # The Cython routines used require contiguous arrays
+    X = np.asarray(X, order="C")
+
+    # Get distance to kth nearest neighbour
+    nbrs = NearestNeighbors(
+        n_neighbors=min_samples,
+        algorithm=algo,
+        leaf_size=leaf_size,
+        metric=metric,
+        metric_params=metric_params,
+        n_jobs=n_jobs,
+        p=None,
+    ).fit(X)
+
+    n_samples = X.shape[0]
+    core_distances = np.empty(n_samples, dtype=np.float64)
+    core_distances[:] = nbrs.kneighbors(X, min_samples)[0][:, -1]
+    dist_metric = DistanceMetric.get_metric(metric, **metric_params)
+
+    # Mutual reachability distance is implicit in mst_linkage_core_vector
+    min_spanning_tree = mst_linkage_core_vector(X, core_distances, dist_metric, alpha)
+
+    return _process_mst(min_spanning_tree)
+
+
+def remap_single_linkage_tree(tree, internal_to_raw, outliers):
+    """
+    Takes an internal single_linkage_tree structure and adds back in a set of points
+    that were initially detected as non-finite and returns that new tree.
+    These points will all be merged into the final node at np.inf distance and
+    considered noise points.
+
+    Parameters
+    ----------
+    tree: single_linkage_tree
+    internal_to_raw: dict
+        a mapping from internal integer index to the raw integer index
+    finite_index: ndarray
+        Boolean array of which entries in the raw data were finite
+    """
+    finite_count = len(internal_to_raw)
+
+    outlier_count = len(outliers)
+    for i, (left, right, distance, size) in enumerate(tree):
+        if left < finite_count:
+            tree[i, 0] = internal_to_raw[left]
+        else:
+            tree[i, 0] = left + outlier_count
+        if right < finite_count:
+            tree[i, 1] = internal_to_raw[right]
+        else:
+            tree[i, 1] = right + outlier_count
+
+    outlier_tree = np.zeros((len(outliers), 4))
+    last_cluster_id = tree[tree.shape[0] - 1][0:2].max()
+    last_cluster_size = tree[tree.shape[0] - 1][3]
+    for i, outlier in enumerate(outliers):
+        outlier_tree[i] = (outlier, last_cluster_id + 1, np.inf, last_cluster_size + 1)
+        last_cluster_id += 1
+        last_cluster_size += 1
+    tree = np.vstack([tree, outlier_tree])
+    return tree
+
+
+def get_finite_row_indices(matrix):
+    """
+    Returns the indices of the purely finite rows of a
+    sparse matrix or dense ndarray
+    """
+    if issparse(matrix):
+        row_indices = np.array(
+            [i for i, row in enumerate(matrix.tolil().data) if np.all(np.isfinite(row))]
+        )
+    else:
+        row_indices = np.where(np.isfinite(matrix).sum(axis=1) == matrix.shape[1])[0]
+    return row_indices
+
+
+@validate_params(
+    {
+        **_PARAM_CONSTRAINTS,
+        "X": ["array-like", "sparse matrix"],
+    }
+)
+def hdbscan(
```

Comment:
I am debating if we should have `hdbscan` be a public function. I think most should call `HDBSCAN().fit_predict` and not use `hdbscan`. `HDBSCAN` handles cases where the input is not all finite.

Reply:
Refactored to remove `hdbscan`. Its functionality can be recovered by calling `HDBSCAN(...).fit(X)` and accessing the necessary attributes directly.